File: bench/cross-language/tree/tree.zig
const std = @import("std");

const Node = struct {
    left: ?*Node,
    right: ?*Node,
    value: i32,
};

var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);

fn buildTree(depth: i32) *Node {
    const node = arena.allocator().create(Node) catch unreachable;
    if (depth <= 0) {
        node.* = .{ .left = null, .right = null, .value = 1 };
        return node;
    }
    node.* = .{
        .left = buildTree(depth - 1),
        .right = buildTree(depth - 1),
        .value = 0,
    };
    return node;
}

fn checkTree(node: *Node) i32 {
    if (node.left == null) {
        return node.value;
    }
    return checkTree(node.left.?) + checkTree(node.right.?);
}

pub fn main() !void {
    const tree = buildTree(18);
    const checksum = checkTree(tree);

    const stdout_file = std.fs.File.stdout();
    var buf: [4096]u8 = undefined;
    var w = stdout_file.writer(&buf);
    try w.interface.print("Checksum: {d}\n", .{checksum});
    try w.interface.flush();

    arena.deinit();
}

File: bench/cross-language/tree/tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    struct Node* left;
    struct Node* right;
    int value;
} Node;

Node* build_tree(int depth) {
    Node* node = (Node*)malloc(sizeof(Node));
    if (depth <= 0) {
        node->left = NULL;
        node->right = NULL;
        node->value = 1;
        return node;
    }
    node->left = build_tree(depth - 1);
    node->right = build_tree(depth - 1);
    node->value = 0;
    return node;
}

int check_tree(Node* node) {
    if (node->left == NULL) {
        return node->value;
    }
    return check_tree(node->left) + check_tree(node->right);
}

void free_tree(Node* node) {
    if (node == NULL) return;
    free_tree(node->left);
    free_tree(node->right);
    free(node);
}

int main() {
    Node* tree = build_tree(18);
    int checksum = check_tree(tree);
    printf("Checksum: %d\n", checksum);
    free_tree(tree);
    return 0;
}
